sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -0,0 +1,488 @@
+      this.maxReadTime = maxReadTime;
+    }
+
+    /**
+     * Creates an {@link UnboundedSource<JmsRecord, ?>} with the configuration in
+     * {@link Read}. Primary use case is unit tests, should not be used in an
+     * application.
+     */
+    @VisibleForTesting
+    UnboundedSource<JmsRecord, JmsCheckpointMark> createSource() {
+      return new UnboundedJmsSource(
+          connectionFactory,
+          queue,
+          topic);
+    }
+
+  }
+
+  private JmsIO() {}
+
+  private static class UnboundedJmsSource extends UnboundedSource<JmsRecord, JmsCheckpointMark> {
+
+    private final ConnectionFactory connectionFactory;
+    private final String queue;
+    private final String topic;
+
+    public UnboundedJmsSource(
+        ConnectionFactory connectionFactory,
+        String queue,
+        String topic) {
+      this.connectionFactory = connectionFactory;
+      this.queue = queue;
+      this.topic = topic;
+    }
+
+    @Override
+    public List<UnboundedJmsSource> generateInitialSplits(
+        int desiredNumSplits, PipelineOptions options) throws Exception {
+      return Collections.singletonList(new UnboundedJmsSource(connectionFactory, queue, topic));
+    }
+
+    @Override
+    public UnboundedJmsReader createReader(PipelineOptions options,
+                                           JmsCheckpointMark checkpointMark) {
+      return new UnboundedJmsReader(this, checkpointMark);
+    }
+
+    @Override
+    public void validate() {
+      if (connectionFactory == null) {
+        throw new IllegalArgumentException("ConnectionFactory is not defined");
+      }
+      if (queue == null && topic == null) {
+        throw new IllegalArgumentException("Either queue or topic is not defined");
+      }
+    }
+
+    @Override
+    public Coder getCheckpointMarkCoder() {
+      return SerializableCoder.of(JmsCheckpointMark.class);
+    }
+
+    @Override
+    public Coder<JmsRecord> getDefaultOutputCoder() {
+      return SerializableCoder.of(JmsRecord.class);
+    }
+
+  }
+
+  private static class UnboundedJmsReader extends UnboundedReader<JmsRecord> {
+
+    private UnboundedJmsSource source;
+    private JmsCheckpointMark checkpointMark;
+    private String destination;
+    private Connection connection;
+    private Session session;
+    private MessageConsumer consumer;
+
+    private JmsRecord currentRecord;
+    private Instant currentTimestamp;
+
+    public UnboundedJmsReader(
+        UnboundedJmsSource source,
+        @Nullable JmsCheckpointMark checkpointMark) {
+      this.source = source;
+      this.checkpointMark = checkpointMark;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      ConnectionFactory connectionFactory = source.connectionFactory;
+      try {
+        this.connection = connectionFactory.createConnection();
+        this.connection.start();
+        this.session = this.connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
```

Comment:
In the latest commit, I'm now using `CLIENT_ACK` and use the checkpoint mark to hold the messages and commit on finalize.
On the other hand, regarding `createSession` without the transaction argument, it depends of the JMS specification version. I would prefer to explicitly define the two arguments to be able to work with "all" JMS version.